File: IPD/ai-rag-medical/src/medrag/indexer.py
import hashlib
import html
import re
import sqlite3
from dataclasses import dataclass
from pathlib import Path
from typing import Iterable

from .config import DEFAULT_DB_PATH, DEFAULT_CHROMA_PATH, DEFAULT_WORKSPACE_ROOT, MATERI_GLOB, EMBEDDING_MODEL
from .models import ChunkRecord, ImageRecord, SourcePage

HEADING_RE = re.compile(r"^(#{1,6})\s+(.+?)\s*$", re.MULTILINE)
IMAGE_RE = re.compile(r"!\[(.*?)\]\((.*?)\)")
WHITESPACE_RE = re.compile(r"\s+")
TABLE_ROW_RE = re.compile(r"^\s*\|.+\|", re.MULTILINE)
TABLE_SEP_RE = re.compile(r"^\s*\|[\s:|-]+\|", re.MULTILINE)
NUMBERED_LIST_RE = re.compile(r"^\s*\d+[.)]\s+", re.MULTILINE)
DOSE_BLOCK_RE = re.compile(r"\d+\s*(?:mg|mcg|µg|unit|ml|g/dl|g/dL|mmol|mEq|IU|gram|tablet|kapsul|ampul|vial)", re.IGNORECASE)

# Section rules for KG edge extraction
SECTION_CATEGORY_MAP = {
    "definisi": "Definisi",
    "pengertian": "Definisi",
    "etiologi": "Etiologi",
    "faktor risiko": "Etiologi",
    "penyebab": "Etiologi",
    "patogenesis": "Patogenesis",
    "patofisiologi": "Patogenesis",
    "mekanisme": "Patogenesis",
    "fisiologi": "Patogenesis",
    "anamnesis": "Anamnesis",
    "anamnesa": "Anamnesis",
    "riwayat": "Anamnesis",
    "manifestasi": "Manifestasi_Klinis",
    "gejala": "Manifestasi_Klinis",
    "klinis": "Manifestasi_Klinis",
    "keluhan": "Manifestasi_Klinis",
    "pemeriksaan fisik": "Pemeriksaan_Fisik",
    "diagnosis": "Diagnosis",
    "pemeriksaan": "Diagnosis",
    "penunjang": "Diagnosis",
    "laboratorium": "Diagnosis",
    "radiologi": "Diagnosis",
    "tata laksana": "Tatalaksana",
    "terapi": "Tatalaksana",
    "pengobatan": "Tatalaksana",
    "obat": "Tatalaksana",
    "farmakologi": "Tatalaksana",
    "dosis": "Tatalaksana",
    "regimen": "Tatalaksana",
    "komplikasi": "Komplikasi",
    "prognosis": "Prognosis",
}


def _checksum(value: str) -> str:
    return hashlib.sha256(value.encode("utf-8")).hexdigest()


def _normalize_text(value: str) -> str:
    decoded = html.unescape(value.replace("&nbsp;", " "))
    return WHITESPACE_RE.sub(" ", decoded).strip()


def discover_source_pages(workspace_root: Path) -> list[SourcePage]:
    pages: list[SourcePage] = []
    for md_path in workspace_root.glob(MATERI_GLOB):
        source_name = md_path.parents[2].name
        page_dir = md_path.parent.name
        page_no = int(page_dir.replace("page-", "")) if page_dir.startswith("page-") else -1
        pages.append(SourcePage(source_name=source_name, page_no=page_no, markdown_path=md_path))
    pages.sort(key=lambda item: (item.source_name, item.page_no))
    return pages


def _split_sections(markdown: str) -> list[tuple[str, str, int, str]]:
    """Split markdown into sections, returning (heading, content, level, parent_heading)."""
    matches = list(HEADING_RE.finditer(markdown))
    if not matches:
        return [("General", markdown, 1, "")]

    sections: list[tuple[str, str, int, str]] = []
    parent_stack: list[tuple[int, str]] = []

    for idx, match in enumerate(matches):
        level = len(match.group(1))
        heading = _normalize_text(match.group(2)) or "General"
        start = match.end()
        end = matches[idx + 1].start() if idx + 1 < len(matches) else len(markdown)
        content = markdown[start:end]

        while parent_stack and parent_stack[-1][0] >= level:
            parent_stack.pop()
        parent_heading = parent_stack[-1][1] if parent_stack else ""
        parent_stack.append((level, heading))

        sections.append((heading, content, level, parent_heading))
    return sections


@dataclass
class _SemanticBlock:
    text: str
    block_type: str  # "prose" | "table" | "list" | "dose_block"


def _split_into_semantic_blocks(text: str) -> list[_SemanticBlock]:
    """Split text into semantic blocks: tables, numbered lists, dose blocks, and prose."""
    lines = text.split("\n")
    blocks: list[_SemanticBlock] = []
    current_lines: list[str] = []
    current_type = "prose"

    def _flush():
        nonlocal current_lines, current_type
        body = "\n".join(current_lines).strip()
        if body:
            blocks.append(_SemanticBlock(text=body, block_type=current_type))
        current_lines = []
        current_type = "prose"

    i = 0
    while i < len(lines):
        line = lines[i]
        is_table = bool(TABLE_ROW_RE.match(line))
        is_numbered = bool(NUMBERED_LIST_RE.match(line))

        if is_table:
            if current_type != "table":
                _flush()
                current_type = "table"
            current_lines.append(line)
        elif is_numbered:
            if current_type != "list":
                _flush()
                current_type = "list"
            current_lines.append(line)
        else:
            if current_type in ("table", "list"):
                if not line.strip():
                    current_lines.append(line)
                    _flush()
                else:
                    _flush()
                    current_type = "prose"
                    current_lines.append(line)
            else:
                current_lines.append(line)
        i += 1

    _flush()

    final_blocks: list[_SemanticBlock] = []
    for block in blocks:
        if block.block_type == "prose" and DOSE_BLOCK_RE.search(block.text):
            block = _SemanticBlock(text=block.text, block_type="dose_block")
        final_blocks.append(block)

    return final_blocks


def _chunk_text(
    value: str,
    max_chars: int = 1800,
    table_max: int = 2500,
    heading: str = "",
    source_name: str = "",
    page_no: int = 0,
) -> list[tuple[str, str]]:
    """Returns list of (chunk_content, content_type) tuples."""
    text = value.strip()
    if not text:
        return []

    metadata_prefix = f"[Konteks: {source_name} - Hal {page_no} | Topik Utama: {heading}]\n"

    semantic_blocks = _split_into_semantic_blocks(text)
    chunks: list[tuple[str, str]] = []

    current_chunk = ""
    current_type = "prose"
    last_paragraph = ""

    def _push_chunk(content: str, ctype: str):
        normalized = _normalize_text(content) if ctype == "prose" else content.strip()
        if normalized:
            chunks.append((metadata_prefix + normalized, ctype))

    for block in semantic_blocks:
        budget = table_max if block.block_type == "table" else max_chars
        max_payload = budget - len(metadata_prefix)

        if block.block_type in ("table", "list", "dose_block"):
            if current_chunk.strip():
                _push_chunk(current_chunk, current_type)
                current_chunk = ""
                last_paragraph = ""

            if len(block.text) <= max_payload:
                _push_chunk(block.text, block.block_type)
            else:
                lines = block.text.split("\n")
                sub = ""
                for line in lines:
                    if len(sub) + len(line) + 1 > max_payload and sub:
                        _push_chunk(sub, block.block_type)
                        sub = ""
                    sub += line + "\n"
                if sub.strip():
                    _push_chunk(sub, block.block_type)
            continue

        max_payload = max_chars - len(metadata_prefix)
        paragraphs = re.split(r"\n\s*\n", block.text)
        current_type = "prose"

        for p in paragraphs:
            if len(current_chunk) + len(p) > max_payload and current_chunk:
                _push_chunk(current_chunk, current_type)
                if len(last_paragraph) < (max_payload * 0.25):
                    current_chunk = last_paragraph + "\n\n"
                else:
                    current_chunk = ""

            if len(p) >= max_payload:
                if current_chunk:
                    _push_chunk(current_chunk, current_type)
                    current_chunk = ""
                lines = p.split("\n")
                sub = ""
                for line in lines:
                    if len(sub) + len(line) < max_payload:
                        sub += line + "\n"
                    else:
                        if sub:
                            _push_chunk(sub, current_type)
                        sub = line + "\n"
                current_chunk = sub + "\n"
                last_paragraph = ""
            else:
                current_chunk += p + "\n\n"
                last_paragraph = p

    if current_chunk.strip():
        _push_chunk(current_chunk, current_type)

    return chunks


def _derive_tags(heading: str, text: str) -> str:
    seeds = f"{heading} {text[:300]}".lower()
    tokens = re.findall(r"[a-zA-Z][a-zA-Z0-9/-]{2,}", seeds)
    unique: list[str] = []
    for token in tokens:
        if token not in unique:
            unique.append(token)
        if len(unique) == 12:
            break
    return " ".join(unique)


def _detect_section_category(heading: str) -> str | None:
    """Heuristic detection of which clinical section a heading belongs to."""
    h_lower = heading.lower()
    for keyword, category in SECTION_CATEGORY_MAP.items():
        if keyword in h_lower:
            return category
    return None


def parse_page(source_page: SourcePage) -> tuple[list[ChunkRecord], list[ImageRecord]]:
    markdown = source_page.markdown_path.read_text(encoding="utf-8", errors="ignore")
    sections = _split_sections(markdown)

    chunk_records: list[ChunkRecord] = []
    image_records: list[ImageRecord] = []
    found_image_refs: set[str] = set()

    for heading, raw_content, level, parent_heading in sections:
        for alt_text, image_ref in IMAGE_RE.findall(raw_content):
            image_abs_path = (source_page.markdown_path.parent / image_ref).resolve()
            nearby = _normalize_text(IMAGE_RE.sub(" ", raw_content))[:300]
            checksum_seed = f"{source_page.markdown_path}|{heading}|{image_ref}|{nearby}"
            found_image_refs.add(image_ref)
            image_records.append(
                ImageRecord(
                    source_name=source_page.source_name,
                    page_no=source_page.page_no,
                    alt_text=_normalize_text(alt_text),
                    image_ref=image_ref,
                    image_abs_path=str(image_abs_path),
                    heading=heading,
                    nearby_text=nearby,
                    markdown_path=str(source_page.markdown_path),
                    checksum=_checksum(checksum_seed),
                )
            )

        cleaned = IMAGE_RE.sub(" ", raw_content)
        section_category = _detect_section_category(heading)
        section_chunks = _chunk_text(
            cleaned,
            heading=heading,
            source_name=source_page.source_name,
            page_no=source_page.page_no,
        )
        total_in_section = len(section_chunks)

        for chunk_idx, (chunk_content, content_type) in enumerate(section_chunks):
            checksum_seed = f"{source_page.markdown_path}|{heading}|{chunk_content}"
            chunk_records.append(
                ChunkRecord(
                    source_name=source_page.source_name,
                    page_no=source_page.page_no,
                    heading=heading,
                    content=chunk_content,
                    disease_tags=_derive_tags(heading, chunk_content),
                    markdown_path=str(source_page.markdown_path),
                    checksum=_checksum(checksum_seed),
                    section_category=section_category or "Ringkasan_Klinis",
                    parent_heading=parent_heading,
                    chunk_index=chunk_idx,
                    total_chunks=total_in_section,
                    heading_level=level,
                    content_type=content_type,
                )
            )

    if not found_image_refs:
        for image_path in sorted(source_page.markdown_path.parent.glob("img-*.*")):
            if image_path.suffix.lower() not in {".jpg", ".jpeg", ".png", ".webp", ".gif"}:
                continue
            checksum_seed = f"{source_page.markdown_path}|fallback|{image_path.name}"
            image_records.append(
                ImageRecord(
                    source_name=source_page.source_name,
                    page_no=source_page.page_no,
                    alt_text=image_path.name,
                    image_ref=image_path.name,
                    image_abs_path=str(image_path.resolve()),
                    heading="Image from page folder",
                    nearby_text="",
                    markdown_path=str(source_page.markdown_path),
                    checksum=_checksum(checksum_seed),
                )
            )

    return chunk_records, image_records


def _init_schema(conn: sqlite3.Connection) -> None:
    conn.executescript(
        """
        PRAGMA journal_mode = WAL;

        DROP TABLE IF EXISTS chunks_fts;
        DROP TABLE IF EXISTS chunks;
        DROP TABLE IF EXISTS images;
        DROP TABLE IF EXISTS graph_edges;

        CREATE TABLE chunks (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            source_name TEXT NOT NULL,
            page_no INTEGER NOT NULL,
            heading TEXT NOT NULL,
            content TEXT NOT NULL,
            disease_tags TEXT NOT NULL,
            markdown_path TEXT NOT NULL,
            checksum TEXT NOT NULL,
            section_category TEXT NOT NULL DEFAULT 'Ringkasan_Klinis',
            parent_heading TEXT NOT NULL DEFAULT '',
            chunk_index INTEGER NOT NULL DEFAULT 0,
            total_chunks INTEGER NOT NULL DEFAULT 1,
            heading_level INTEGER NOT NULL DEFAULT 2,
            content_type TEXT NOT NULL DEFAULT 'prose'
        );

        CREATE VIRTUAL TABLE chunks_fts USING fts5(
            content,
            disease_tags,
            source_name,
            heading,
            content='chunks',
            content_rowid='id'
        );

        CREATE TABLE images (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            source_name TEXT NOT NULL,
            page_no INTEGER NOT NULL,
            alt_text TEXT,
            image_ref TEXT NOT NULL,
            image_abs_path TEXT NOT NULL,
            heading TEXT NOT NULL,
            nearby_text TEXT NOT NULL,
            markdown_path TEXT NOT NULL,
            checksum TEXT NOT NULL
        );

        CREATE TABLE graph_edges (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            source_disease TEXT NOT NULL,
            relation TEXT NOT NULL,
            target_node TEXT NOT NULL,
            target_type TEXT NOT NULL,
            source_name TEXT NOT NULL,
            page_no INTEGER NOT NULL
        );
        """
    )


def _insert_records(conn: sqlite3.Connection, chunks: list[ChunkRecord], images: list[ImageRecord]) -> None:
    conn.executemany(
        """
        INSERT INTO chunks (source_name, page_no, heading, content, disease_tags, markdown_path, checksum,
                            section_category, parent_heading, chunk_index, total_chunks, heading_level, content_type)
        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
        """,
        [
            (
                record.source_name,
                record.page_no,
                record.heading,
                record.content,
                record.disease_tags,
                record.markdown_path,
                record.checksum,
                record.section_category,
                record.parent_heading,
                record.chunk_index,
                record.total_chunks,
                record.heading_level,
                record.content_type,
            )
            for record in chunks
        ],
    )

    conn.execute(
        """
        INSERT INTO chunks_fts(rowid, content, disease_tags, source_name, heading)
        SELECT id, content, disease_tags, source_name, heading FROM chunks
        """
    )

    conn.executemany(
        """
        INSERT INTO images (source_name, page_no, alt_text, image_ref, image_abs_path, heading, nearby_text, markdown_path, checksum)
        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
        """,
        [
            (
                record.source_name,
                record.page_no,
                record.alt_text,
                record.image_ref,
                record.image_abs_path,
                record.heading,
                record.nearby_text,
                record.markdown_path,
                record.checksum,
            )
            for record in images
        ],
    )


def _insert_graph_edges(conn: sqlite3.Connection, chunks: list[ChunkRecord]) -> None:
    """Extract heuristic knowledge graph edges from chunk headings."""
    edges = []
    for chunk in chunks:
        disease_node = chunk.source_name
        category = chunk.section_category
        # Extract key terms from heading as target nodes
        heading_words = re.findall(r"[A-Za-zÀ-ÿ][A-Za-zÀ-ÿ0-9/ ]{2,}", chunk.heading)
        for term in heading_words[:3]:
            term = term.strip()
            if term and term.lower() != disease_node.lower():
                edges.append((disease_node, category, term, "concept", chunk.source_name, chunk.page_no))

    if edges:
        conn.executemany(
            "INSERT INTO graph_edges (source_disease, relation, target_node, target_type, source_name, page_no) VALUES (?, ?, ?, ?, ?, ?)",
            edges,
        )


def _build_vector_index(chunks: list[ChunkRecord], chroma_path: Path) -> None:
    """Build ChromaDB vector index from chunks using sentence-transformers."""
    try:
        import chromadb
        from sentence_transformers import SentenceTransformer
    except ImportError:
        print("[WARN] chromadb or sentence-transformers not installed. Skipping vector index.")
        return

    print("[INFO] Building vector index (this may take a while on first run)...")
    chroma_path.mkdir(parents=True, exist_ok=True)

    client = chromadb.PersistentClient(path=str(chroma_path))

    # Delete existing collection to rebuild fresh
    try:
        client.delete_collection("medrag_chunks")
    except Exception:
        pass

    collection = client.create_collection(
        name="medrag_chunks",
        metadata={"hnsw:space": "cosine"},
    )

    model = SentenceTransformer(EMBEDDING_MODEL)

    batch_size = 64
    for i in range(0, len(chunks), batch_size):
        batch = chunks[i : i + batch_size]
        texts = [f"passage: {c.heading}: {c.content}" for c in batch]
        embeddings = model.encode(texts, show_progress_bar=False).tolist()
        ids = [f"chunk_{i + j}" for j in range(len(batch))]
        metadatas = [
            {
                "source_name": c.source_name,
                "page_no": c.page_no,
                "heading": c.heading,
                "section_category": c.section_category,
                "disease_tags": c.disease_tags,
                "content": c.content,
                "parent_heading": c.parent_heading,
                "content_type": c.content_type,
                "chunk_index": c.chunk_index,
            }
            for c in batch
        ]
        collection.add(ids=ids, embeddings=embeddings, documents=texts, metadatas=metadatas)

    print(f"[INFO] Vector index built: {len(chunks)} chunks embedded.")


def build_index(
    workspace_root: Path | None = None,
    db_path: Path | None = None,
    chroma_path: Path | None = None,
    skip_vector: bool = False,
) -> dict[str, int]:
    workspace = workspace_root or DEFAULT_WORKSPACE_ROOT
    database = db_path or DEFAULT_DB_PATH
    chroma = chroma_path or DEFAULT_CHROMA_PATH
    database.parent.mkdir(parents=True, exist_ok=True)

    source_pages = discover_source_pages(workspace)

    all_chunks: list[ChunkRecord] = []
    all_images: list[ImageRecord] = []
    for page in source_pages:
        chunks, images = parse_page(page)
        all_chunks.extend(chunks)
        all_images.extend(images)

    conn = sqlite3.connect(database)
    try:
        _init_schema(conn)
        _insert_records(conn, all_chunks, all_images)
        _insert_graph_edges(conn, all_chunks)
        conn.commit()
    finally:
        conn.close()

    if not skip_vector:
        _build_vector_index(all_chunks, chroma)

    return {
        "source_pages": len(source_pages),
        "chunks": len(all_chunks),
        "images": len(all_images),
    }

File: IPD/ai-rag-medical/worker/src/indexer.ts
import type { Env } from "./types";

export const WORKERS_EMBEDDING_MODEL = "@cf/baai/bge-base-en-v1.5";

const HEADING_RE = /^(#{1,6})\s+(.+?)\s*$/gm;
const IMAGE_RE = /!\[(.*?)\]\((.*?)\)/g;
const WHITESPACE_RE = /\s+/g;
const TABLE_ROW_RE = /^\s*\|.+\|/m;
const NUMBERED_LIST_RE = /^\s*\d+[.)]\s+/m;
const DOSE_BLOCK_RE = /\d+\s*(?:mg|mcg|µg|unit|ml|g\/dl|g\/dL|mmol|mEq|IU|gram|tablet|kapsul|ampul|vial)/i;
const TOKEN_RE = /[a-zA-Z][a-zA-Z0-9/-]{2,}/g;

const SECTION_CATEGORY_MAP: Array<[string, string]> = [
  ["definisi", "Definisi"],
  ["pengertian", "Definisi"],
  ["batasan", "Definisi"],
  ["deskripsi", "Definisi"],
  ["etiologi", "Etiologi"],
  ["faktor risiko", "Etiologi"],
  ["penyebab", "Etiologi"],
  ["kausa", "Etiologi"],
  ["predisposisi", "Etiologi"],
  ["epidemiologi", "Etiologi"],
  ["insidensi", "Etiologi"],
  ["prevalensi", "Etiologi"],
  ["patogenesis", "Patogenesis"],
  ["patofisiologi", "Patogenesis"],
  ["mekanisme", "Patogenesis"],
  ["fisiologi", "Patogenesis"],
  ["imunopatologi", "Patogenesis"],
  ["anamnesis", "Anamnesis"],
  ["anamnesa", "Anamnesis"],
  ["riwayat", "Anamnesis"],
  ["keluhan utama", "Anamnesis"],
  ["manifestasi", "Manifestasi_Klinis"],
  ["gejala", "Manifestasi_Klinis"],
  ["klinis", "Manifestasi_Klinis"],
  ["keluhan", "Manifestasi_Klinis"],
  ["gambaran klinis", "Manifestasi_Klinis"],
  ["tanda klinis", "Manifestasi_Klinis"],
  ["simptom", "Manifestasi_Klinis"],
  ["sindrom", "Manifestasi_Klinis"],
  ["presentasi", "Manifestasi_Klinis"],
  ["gejala klinis", "Manifestasi_Klinis"],
  ["pemeriksaan fisik", "Pemeriksaan_Fisik"],
  ["physical exam", "Pemeriksaan_Fisik"],
  ["diagnosis", "Diagnosis"],
  ["pemeriksaan", "Diagnosis"],
  ["penunjang", "Diagnosis"],
  ["laboratorium", "Diagnosis"],
  ["radiologi", "Diagnosis"],
  ["kriteria", "Diagnosis"],
  ["gold standard", "Diagnosis"],
  ["baku emas", "Diagnosis"],
  ["staging", "Diagnosis"],
  ["klasifikasi", "Diagnosis"],
  ["grading", "Diagnosis"],
  ["tata laksana", "Tatalaksana"],
  ["tatalaksana", "Tatalaksana"],
  ["terapi", "Tatalaksana"],
  ["pengobatan", "Tatalaksana"],
  ["penatalaksanaan", "Tatalaksana"],
  ["farmakologi", "Tatalaksana"],
  ["obat", "Tatalaksana"],
  ["dosis", "Tatalaksana"],
  ["regimen", "Tatalaksana"],
  ["medikamentosa", "Tatalaksana"],
  ["non-farmakologi", "Tatalaksana"],
  ["nonfarmakologi", "Tatalaksana"],
  ["intervensi", "Tatalaksana"],
  ["protokol", "Tatalaksana"],
  ["penanganan", "Tatalaksana"],
  ["manajemen", "Tatalaksana"],
  ["komplikasi", "Komplikasi"],
  ["penyulit", "Komplikasi"],
  ["prognosis", "Prognosis"],
  ["luaran", "Prognosis"],
  ["mortalitas", "Prognosis"],
  ["survival", "Prognosis"],
  ["angka kematian", "Prognosis"],
  ["pencegahan", "Prognosis"],
];

export type SourcePageInput = {
  source_name: string;
  page_no: number;
  markdown: string;
  stase_slug?: string;
  markdown_path?: string;
};

export type ParsedChunk = {
  source_name: string;
  page_no: number;
  heading: string;
  content: string;
  disease_tags: string;
  markdown_path: string;
  checksum: string;
  section_category: string;
  parent_heading: string;
  chunk_index: number;
  total_chunks: number;
  heading_level: number;
  content_type: string;
  stase_slug: string;
};

export type ParsedImage = {
  source_name: string;
  page_no: number;
  alt_text: string;
  image_ref: string;
  image_abs_path: string;
  heading: string;
  nearby_text: string;
  markdown_path: string;
  checksum: string;
  stase_slug: string;
};

export type ParsedGraphEdge = {
  source_disease: string;
  relation: string;
  target_node: string;
  target_type: string;
  source_name: string;
  page_no: number;
  stase_slug: string;
};

export type ParsedPage = {
  chunks: ParsedChunk[];
  images: ParsedImage[];
  graph_edges: ParsedGraphEdge[];
};

function checksum(value: string): string {
  return value;
}

function normalizeText(value: string): string {
  return value.replace(/&nbsp;/g, " ").replace(WHITESPACE_RE, " ").trim();
}

function detectSectionCategory(heading: string): string {
  const lower = heading.toLowerCase();
  for (const [keyword, category] of [...SECTION_CATEGORY_MAP].sort((a, b) => b[0].length - a[0].length)) {
    if (lower.includes(keyword)) return category;
  }
  return "Ringkasan_Klinis";
}

function deriveDiseaseTags(heading: string, text: string): string {
  const seeds = `${heading} ${text.slice(0, 300)}`.toLowerCase();
  const tokens = seeds.match(TOKEN_RE) ?? [];
  const unique: string[] = [];
  for (const token of tokens) {
    if (!unique.includes(token)) unique.push(token);
    if (unique.length === 12) break;
  }
  return unique.join(" ");
}

function splitSections(markdown: string): Array<{ heading: string; content: string; level: number; parent_heading: string }> {
  const matches = [...markdown.matchAll(HEADING_RE)];
  if (matches.length === 0) {
    return [{ heading: "General", content: markdown, level: 1, parent_heading: "" }];
  }

  const sections: Array<{ heading: string; content: string; level: number; parent_heading: string }> = [];
  const parentStack: Array<{ level: number; heading: string }> = [];

  matches.forEach((match, index) => {
    const level = match[1].length;
    const heading = normalizeText(match[2] ?? "") || "General";
    const start = (match.index ?? 0) + match[0].length;
    const end = index + 1 < matches.length ? (matches[index + 1].index ?? markdown.length) : markdown.length;
    const content = markdown.slice(start, end);

    while (parentStack.length > 0 && parentStack[parentStack.length - 1].level >= level) {
      parentStack.pop();
    }
    const parentHeading = parentStack[parentStack.length - 1]?.heading ?? "";
    parentStack.push({ level, heading });
    sections.push({ heading, content, level, parent_heading: parentHeading });
  });

  return sections;
}

function splitIntoSemanticBlocks(text: string): Array<{ text: string; block_type: "prose" | "table" | "list" | "dose_block" }> {
  const lines = text.split("\n");
  const blocks: Array<{ text: string; block_type: "prose" | "table" | "list" | "dose_block" }> = [];
  let currentLines: string[] = [];
  let currentType: "prose" | "table" | "list" = "prose";

  const flush = () => {
    const body = currentLines.join("\n").trim();
    if (body) blocks.push({ text: body, block_type: currentType });
    currentLines = [];
    currentType = "prose";
  };

  for (const line of lines) {
    const isTable = TABLE_ROW_RE.test(line);
    const isNumbered = NUMBERED_LIST_RE.test(line);

    if (isTable) {
      if (currentType !== "table") {
        flush();
        currentType = "table";
      }
      currentLines.push(line);
      continue;
    }

    if (isNumbered) {
      if (currentType !== "list") {
        flush();
        currentType = "list";
      }
      currentLines.push(line);
      continue;
    }

    if (currentType === "table" || currentType === "list") {
      if (!line.trim()) {
        currentLines.push(line);
        flush();
      } else {
        flush();
        currentLines.push(line);
        currentType = "prose";
      }
      continue;
    }

    currentLines.push(line);
  }

  flush();

  return blocks.map((block) => {
    if (block.block_type === "prose" && DOSE_BLOCK_RE.test(block.text)) {
      return { ...block, block_type: "dose_block" as const };
    }
    return block;
  });
}

function chunkText(
  value: string,
  maxChars = 1800,
  tableMax = 2500,
  heading = "",
  sourceName = "",
  pageNo = 0,
): Array<{ chunk_content: string; content_type: string }> {
  const text = value.trim();
  if (!text) return [];

  const metadataPrefix = `[Konteks: ${sourceName} - Hal ${pageNo} | Topik Utama: ${heading}]\n`;
  const blocks = splitIntoSemanticBlocks(text);
  const chunks: Array<{ chunk_content: string; content_type: string }> = [];

  let currentChunk = "";
  let currentType = "prose";
  let lastParagraph = "";

  const pushChunk = (content: string, contentType: string) => {
    const normalized = contentType === "prose" ? normalizeText(content) : content.trim();
    if (normalized) {
      chunks.push({ chunk_content: metadataPrefix + normalized, content_type: contentType });
    }
  };

  for (const block of blocks) {
    const budget = block.block_type === "table" ? tableMax : maxChars;
    const maxPayload = budget - metadataPrefix.length;

    if (block.block_type === "table" || block.block_type === "list" || block.block_type === "dose_block") {
      if (currentChunk.trim()) {
        pushChunk(currentChunk, currentType);
        currentChunk = "";
        lastParagraph = "";
      }

      if (block.text.length <= maxPayload) {
        pushChunk(block.text, block.block_type);
      } else {
        const lines = block.text.split("\n");
        let sub = "";
        for (const line of lines) {
          if (sub.length + line.length + 1 > maxPayload && sub) {
            pushChunk(sub, block.block_type);
            sub = "";
          }
          sub += `${line}\n`;
        }
        if (sub.trim()) pushChunk(sub, block.block_type);
      }
      continue;
    }

    const proseMaxPayload = maxChars - metadataPrefix.length;
    const paragraphs = block.text.split(/\n\s*\n/);
    currentType = "prose";

    for (const paragraph of paragraphs) {
      if (currentChunk.length + paragraph.length > proseMaxPayload && currentChunk) {
        pushChunk(currentChunk, currentType);
        currentChunk = lastParagraph.length < proseMaxPayload * 0.25 ? `${lastParagraph}\n\n` : "";
      }

      if (paragraph.length >= proseMaxPayload) {
        if (currentChunk) {
          pushChunk(currentChunk, currentType);
          currentChunk = "";
        }
        const lines = paragraph.split("\n");
        let sub = "";
        for (const line of lines) {
          if (sub.length + line.length < proseMaxPayload) {
            sub += `${line}\n`;
          } else {
            if (sub) pushChunk(sub, currentType);
            sub = `${line}\n`;
          }
        }
        currentChunk = `${sub}\n`;
        lastParagraph = "";
        continue;
      }

      currentChunk += `${paragraph}\n\n`;
      lastParagraph = paragraph;
    }
  }

  if (currentChunk.trim()) {
    pushChunk(currentChunk, currentType);
  }

  return chunks;
}

function extractGraphEdgesFromChunks(chunks: ParsedChunk[]): ParsedGraphEdge[] {
  const edges: ParsedGraphEdge[] = [];
  for (const chunk of chunks) {
    const headingWords = chunk.heading.match(/[A-Za-zÀ-ÿ][A-Za-zÀ-ÿ0-9/ ]{2,}/g) ?? [];
    for (const term of headingWords.slice(0, 3)) {
      const clean = term.trim();
      if (clean && clean.toLowerCase() !== chunk.source_name.toLowerCase()) {
        edges.push({
          source_disease: chunk.source_name,
          relation: chunk.section_category,
          target_node: clean,
          target_type: "concept",
          source_name: chunk.source_name,
          page_no: chunk.page_no,
          stase_slug: chunk.stase_slug,
        });
      }
    }
  }
  return edges;
}

export function parsePageForIndexing(page: SourcePageInput): ParsedPage {
  const markdown = page.markdown.replace(/\r\n/g, "\n");
  const sections = splitSections(markdown);
  const markdownPath = page.markdown_path ?? `page-${page.page_no}/markdown.md`;
  const staseSlug = page.stase_slug ?? "ipd";

  const chunks: ParsedChunk[] = [];
  const images: ParsedImage[] = [];
  const foundImageRefs = new Set<string>();

  for (const section of sections) {
    const rawMatches = [...section.content.matchAll(IMAGE_RE)];
    for (const match of rawMatches) {
      const altText = normalizeText(match[1] ?? "");
      const imageRef = (match[2] ?? "").trim();
      if (!imageRef) continue;
      foundImageRefs.add(imageRef);
      const nearby = normalizeText(section.content.replace(IMAGE_RE, " ")).slice(0, 300);
      images.push({
        source_name: page.source_name,
        page_no: page.page_no,
        alt_text: altText,
        image_ref: imageRef,
        image_abs_path: imageRef,
        heading: section.heading,
        nearby_text: nearby,
        markdown_path: markdownPath,
        checksum: checksum(`${markdownPath}|${section.heading}|${imageRef}|${nearby}`),
        stase_slug: staseSlug,
      });
    }

    const cleaned = section.content.replace(IMAGE_RE, " ");
    const sectionCategory = detectSectionCategory(section.heading);
    const sectionChunks = chunkText(cleaned, 1800, 2500, section.heading, page.source_name, page.page_no);

    sectionChunks.forEach((chunk, chunkIndex) => {
      chunks.push({
        source_name: page.source_name,
        page_no: page.page_no,
        heading: section.heading,
        content: chunk.chunk_content,
        disease_tags: deriveDiseaseTags(section.heading, chunk.chunk_content),
        markdown_path: markdownPath,
        checksum: checksum(`${markdownPath}|${section.heading}|${chunk.chunk_content}`),
        section_category: sectionCategory,
        parent_heading: section.parent_heading,
        chunk_index: chunkIndex,
        total_chunks: sectionChunks.length,
        heading_level: section.level,
        content_type: chunk.content_type,
        stase_slug: staseSlug,
      });
    });
  }

  if (!foundImageRefs.size) {
    // Cloud uploads do not have filesystem access, so there is no safe fallback image discovery.
  }

  return {
    chunks,
    images,
    graph_edges: extractGraphEdgesFromChunks(chunks),
  };
}

export async function embedTexts(env: Env, texts: string[]): Promise<number[][]> {
  if (texts.length === 0) return [];
  const result = (await env.AI.run(WORKERS_EMBEDDING_MODEL, {
    text: texts,
  })) as unknown as { data?: number[][]; embedding?: number[][] };
  const vectors = result.data ?? result.embedding ?? [];
  return vectors.filter((row): row is number[] => Array.isArray(row));
}
